src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,551 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Gamma-Poisson reach curve model for underreported counts."""
+
+from absl import logging
+import numpy as np
+import scipy.stats
+from typing import List
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+# The following constant defines the highest frequency that will be
+# used for parameter estimation.
+MAXIMUM_COMPUTATIONAL_FREQUENCY = 1000
+
+
+class KInflatedGammaPoissonDistribution:
+    """k-inflated Gamma-Poisson distribution with specified parameters.
+
+    The k-inflated Gamma-Poisson distribution is like the Gamma-Poisson
+    distribution except that the first k values of the PMF are allowed to
+    be arbitrary.
+
+    It is given by k+2 parameters: alpha, beta, a_1, a_2, ..., a_k.
+    We assume that each a_i >= 0, and that a_1 + a_2 + ... + a_k < 1.
+
+    Let GP(n | alpha, beta) be the PMF of the shifted Gamma-Poisson distribution
+    with parameters alpha, beta.  Also, define s = a_1 + a_2 + ... + a_k,
+    and define t = \sum_{i=1}^k GP(i | alpha, beta).  Then, the PMF of the
+    k-inflated Gamma-Poisson distribution is given as
+
+     kGP(n | alpha, beta, a_1, ..., a_k)
+          = a_n if 0 < n <= k,
+            (1-s) / (1-t) GP(n | alpha, beta) if n > k.
+    """
+
+    def __init__(self, alpha, beta, a):
+        """k-Inflated Gamma-Poisson with parameters alpha, beta, N, a. """
+        if sum(a) > 1.0:
+            raise ValueError("Inflation values sum to a value greater than 1.")
+
+        self._alpha = alpha
+        self._beta = beta
+        self._a = a
+
+        n = len(a)
+        s = np.sum(a)
+        t = np.sum(self._gamma_poisson_pmf(np.arange(1, n + 1), alpha, beta))
+        u = self._gamma_poisson_pmf(
+            np.arange(n + 1, MAXIMUM_COMPUTATIONAL_FREQUENCY), alpha, beta
+        )
+        pmf_tail = ((1 - s) / (1 - t)) * u
+        # self._pmf[n] = kGP(n | alpha, beta, a_1, ..., a_k)
+        self._pmf = np.array([0.0] + list(a) + list(pmf_tail))
```

Comment:
Why including 0?  This was not mentioned in the model description.

Reply:
Done.